### PR TITLE
[Diffusion] Revamp Rollout Log-Prob Support with SDE/CPS for RL Post-Training

### PR DESCRIPTION
**Pull request by Rockdu**
# [Diffusion] Revamp Rollout Log-Prob Support with SDE/CPS for RL Post-Training

### Motivation

RL-based post-training of diffusion models (e.g., FlowGRPO) requires computing per-step log-probabilities along the denoising trajectory. This PR adds a modular rollout log-prob engine to SGLang-D that supports three denoising strategies and is compatible with Sequence Parallelism (SP).

### Design Principles

- **Stateless scheduler**: All intermediate rollout state lives in a `RolloutSessionData` dataclass attached to the batch object (`batch._rollout_session_data`). The scheduler itself stores nothing, making it safe for concurrent use.
- **Composable mixins**: Rollout logic is encapsulated in `SchedulerRLMixin` (core) and `SchedulerRLDebugMixin` (debug tensor collection), mixed into `FlowMatchEulerDiscreteScheduler` alongside the existing `SchedulerMixin` and `ConfigMixin`.
- **Opt-in activation**: Rollout is triggered by `SamplingParams.rollout = True`. When disabled, no additional code paths execute and the scheduler step is unchanged.

### Module Structure

```
runtime/post_training/
  rl_dataclasses.py            # RolloutSessionData, RolloutDebugTensors, RolloutTrajectoryData
  scheduler_rl_mixin.py        # SchedulerRLMixin  (core log-prob engine)
  scheduler_rl_debug_mixin.py  # SchedulerRLDebugMixin  (debug tensor accumulation)
```

### Data Flow

```
prepare_rollout(batch, pipeline_config)
  --> creates RolloutSessionData on batch

for each denoising step:
  scheduler.step(batch=batch, ...)
    --> flow_sde_sampling(batch, model_output, sample, sigma, sigma_next, generator)
        --> computes prev_sample, log_prob_local_sum, local_elem_count
    --> append_local_rollout_log_probs(batch, ...)

after denoising loop:
  collect_rollout_log_probs(batch)      --> batch.rollout_trajectory_data.rollout_log_probs
  collect_rollout_debug_tensors(batch)  --> batch.rollout_trajectory_data.rollout_debug_tensors  (optional)
  release_rollout_resources(batch)      --> batch._rollout_session_data = None
```

---

### Supported Rollout Strategies

| Strategy | `rollout_sde_type` | Description                                                  |
| -------- | ------------------ | ------------------------------------------------------------ |
| SDE      | `"sde"`            | Standard stochastic differential equation with Gaussian diffusion noise |
| CPS      | `"cps"`            | Coefficients-Preserving Sampling https://arxiv.org/pdf/2509.05952 |
| ODE      | `"ode"`            | Deterministic ODE step (zero variance, for alignment verification) |

### API Parameters (`SamplingParams`)

| Parameter                   | Type    | Default | Description                                                  |
| --------------------------- | ------- | ------- | ------------------------------------------------------------ |
| `rollout`                   | `bool`  | `False` | Enable rollout log-prob computation                          |
| `rollout_sde_type`          | `str`   | `"sde"` | Step strategy: `"sde"`, `"cps"`, or `"ode"`                  |
| `rollout_noise_level`       | `float` | `0.7`   | Noise level for SDE/CPS                                      |
| `rollout_log_prob_no_const` | `bool`  | `False` | Omit constant terms in log-prob (common for RL loss)         |
| `rollout_debug_mode`        | `bool`  | `False` | Return intermediate tensors (variance noise, mean, std, model output) |

### Output

When `rollout=True`, the batch's `rollout_trajectory_data` (type `RolloutTrajectoryData`) is populated:

- `rollout_log_probs`: `Tensor [B, T]` -- per-step log-probabilities, reduced as `global_sum / global_count` across SP ranks.
- `rollout_debug_tensors` (when `rollout_debug_mode=True`): `RolloutDebugTensors` containing per-step `[B, T, ...]` tensors for `variance_noise`, `prev_sample_mean`, `noise_std_dev`, and `model_output`.

These are available through the OpenAI-compatible image/video API via the existing extra-fields mechanism.

### Reliability Testing

Three levels of testing validate correctness:

1. **Unit tests** (`test_scheduler_rollout_ode_unit.py`):
   - ODE step determinism: confirms no variance noise is sampled.
   - Debug tensor shapes: verifies `noise_std_dev` is zero and shapes are consistent.
   - **SDE/CPS FlowGRPO alignment**: each single step is compared against a standalone FlowGRPO reference implementation across 4 seeds. Max absolute difference <= 1e-6 for `prev_sample`, `prev_sample_mean`, and `noise_std_dev`.

2. **ODE-vs-non-rollout end-to-end** (Z-Image-Turbo, 1024x1024, seed=42): ODE rollout produces **bit-identical** output images to the standard non-rollout denoising path across all parallelism configs (TP1/SP2, TP2/SP1, TP1/SP1+CFGP).

3. **SDE/CPS parallel consistency** (Z-Image-Turbo, 1024x1024, seed=42): Intermediate tensors (variance noise, noise std, prev sample mean, model output) are compared across parallelism configs against a single-GPU reference, with all steps and first-step metrics reported.

---

### 3.1 ODE vs Non-Rollout (Bit-Exact Alignment)

Model: `Tongyi-MAI/Z-Image-Turbo`, seed=42, 1024x1024, 2 GPUs.

| Config       |   TP |   SP | CFGP Parallel | Exact Match | Max Abs Diff |  MSE | Cosine Sim |
| ------------ | ---: | ---: | ------------- | ----------- | -----------: | ---: | ---------: |
| TP1 SP2      |    1 |    2 | No            | **True**    |          0.0 |  0.0 |        1.0 |
| TP2 SP1      |    2 |    1 | No            | **True**    |          0.0 |  0.0 |        1.0 |
| TP1 SP1 CFGP |    1 |    1 | Yes           | **True**    |          0.0 |  0.0 |        1.0 |

All configs produce **pixel-identical** images, confirming the ODE path introduces zero numerical divergence from the standard denoising pipeline.

### 3.2 SDE/CPS Parallel Consistency

Model: `Tongyi-MAI/Z-Image-Turbo`, seed=42, noise_level=0.5, 9 steps, 2 GPUs.
Reference: single-GPU (TP1, SP1, no CFGP parallel).

**SDE mode -- key metrics (all steps)**:

| Config       | Tensor           | Max Abs Diff | Mean MSE | Min Cosine | Mean Cosine |
| ------------ | ---------------- | -----------: | -------: | ---------: | ----------: |
| TP1 SP2      | variance_noise   |            0 |        0 |        1.0 |         1.0 |
| TP1 SP2      | prev_sample_mean |            0 |        0 |        1.0 |         1.0 |
| TP1 SP2      | noise_std_dev    |            0 |        0 |        1.0 |         1.0 |
| TP2 SP1      | variance_noise   |            0 |        0 |        1.0 |         1.0 |
| TP2 SP1      | prev_sample_mean |         2.17 |  3.68e-3 |     0.9978 |      0.9990 |
| TP2 SP1      | noise_std_dev    |            0 |        0 |        1.0 |         1.0 |
| TP1 SP1 CFGP | variance_noise   |            0 |        0 |        1.0 |         1.0 |
| TP1 SP1 CFGP | prev_sample_mean |         1.45 |  5.97e-4 |     0.9996 |      0.9998 |
| TP1 SP1 CFGP | noise_std_dev    |            0 |        0 |        1.0 |         1.0 |

**CPS mode -- key metrics (all steps)**:

| Config       | Tensor           | Max Abs Diff | Mean MSE | Min Cosine | Mean Cosine |
| ------------ | ---------------- | -----------: | -------: | ---------: | ----------: |
| TP1 SP2      | variance_noise   |            0 |        0 |        1.0 |         1.0 |
| TP1 SP2      | prev_sample_mean |            0 |        0 |        1.0 |         1.0 |
| TP2 SP1      | variance_noise   |            0 |        0 |        1.0 |         1.0 |
| TP2 SP1      | prev_sample_mean |         2.32 |  5.12e-3 |     0.9961 |      0.9980 |
| TP1 SP1 CFGP | variance_noise   |            0 |        0 |        1.0 |         1.0 |
| TP1 SP1 CFGP | prev_sample_mean |         2.10 |  9.20e-4 |     0.9992 |      0.9997 |

**Key observations across both modes**:

- `variance_noise` and `noise_std_dev` are **bit-exact** across all configs -- the SP-aware noise generation correctly reproduces the single-GPU noise stream.
- `prev_sample_mean` and `model_output` show small accumulated differences in TP2/SP1 and CFGP-parallel configs, caused by non-deterministic floating-point reduction order in the transformer (not the rollout engine). First-step differences are negligible (cosine > 0.9999).
- SP2 (sequence parallel on the latent axis) is **bit-exact** with the single-GPU reference on all tensors.

---

## 4. Summary

- The rollout log-prob engine correctly implements SDE, CPS, and ODE strategies as validated by unit tests aligned with FlowGRPO reference formulas.
- **ODE mode produces bit-identical outputs** to the standard non-rollout path across all parallelism configurations, confirming zero regression on existing functionality.
- **Noise generation is bit-exact** under SP -- the `shard_latents_for_sp`-based noise generation guarantees identical random streams regardless of parallelism.
- Small `prev_sample_mean` divergences under TP2/CFGP-parallel are expected floating-point non-determinism from the DiT forward pass, not the rollout engine. These accumulate over steps but remain within cosine similarity > 0.996.
- All rollout state is batch-scoped via `RolloutSessionData` -- the scheduler is fully stateless and safe for multi-request serving.

### CODE REVIEW

File: python/sglang/multimodal_gen/configs/sample/sampling_params.py [hunk 1]
```diff
@@ -803,6 +820,32 @@ def add_argument(*name_or_flags, **kwargs):
             action="store_true",
             help="Whether to return the trajectory",
         )
+        add_argument(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/sample/sampling_params.py]: could we use a dedicated parser (for RL), and add arguments of the parser here

Comment (Rockdu) [reply python/sglang/multimodal_gen/configs/sample/sampling_params.py]: Sure! sounds reasonable since we have many args here exclusively work for RL

File: python/sglang/multimodal_gen/configs/sample/sampling_params.py [hunk 2]
```diff
@@ -536,7 +546,7 @@ def from_user_sampling_params_args(
             sampling_params = SamplingParams.from_pretrained(
                 model_path, backend=server_args.backend, model_id=server_args.model_id
             )
-        except (AttributeError, ValueError) as e:
+        except (AttributeError, ValueError):
             # Handle safetensors files or other cases where model_index.json is not available
             # Use appropriate SamplingParams based on pipeline_class_name from registry
             if os.path.isfile(model_path) and model_path.endswith(".safetensors"):
```
Comment (Rockdu) [inline python/sglang/multimodal_gen/configs/sample/sampling_params.py:569]: Remove these mistakenly edited lines by local lint

File: python/sglang/multimodal_gen/runtime/models/schedulers/scheduling_flow_match_euler_discrete.py [hunk 3]
```diff
@@ -447,6 +452,7 @@ def step(
         s_noise: float = 1.0,
         generator: torch.Generator | None = None,
         per_token_timesteps: torch.Tensor | None = None,
+        batch=None,
```
Comment (Rockdu) [inline python/sglang/multimodal_gen/runtime/models/schedulers/scheduling_flow_match_euler_discrete.py:455]: This edit of interface is only for RL, absorb into mixin preparation

File: python/sglang/multimodal_gen/runtime/models/schedulers/scheduling_flow_match_euler_discrete.py [hunk 4]
```diff
@@ -516,12 +522,22 @@ def step(
             next_sigma = sigma_next
             dt = sigma_next - sigma
 
-        if self.config.stochastic_sampling:
-            x0 = sample - current_sigma * model_output
-            noise = torch.randn_like(sample)
-            prev_sample = (1.0 - next_sigma) * x0 + next_sigma * noise
+        if batch is not None and self.already_prepared_rollout(batch):
```
Comment (Rockdu) [inline python/sglang/multimodal_gen/runtime/models/schedulers/scheduling_flow_match_euler_discrete.py]: Also try to absorb these into mixin

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 5]
```diff
@@ -557,10 +600,12 @@ def _prepare_denoising_loop(self, batch: Req, server_args: ServerArgs):
         else:
             self._maybe_enable_cache_dit(cache_dit_num_inference_steps, batch)
 
+        self._maybe_prepare_rollout(batch)
+
         # Prepare extra step kwargs for scheduler
         extra_step_kwargs = self.prepare_extra_func_kwargs(
             self.scheduler.step,
-            {"generator": batch.generator, "eta": batch.eta},
+            {"generator": batch.generator, "eta": batch.eta, "batch": batch},
```
Comment (Rockdu) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:615]: This logic should not affect scheduler interface

File: python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py [hunk 6]
```diff
@@ -333,8 +337,9 @@ class OutputBatch:
     output: torch.Tensor | None = None
     audio: torch.Tensor | None = None
     audio_sample_rate: int | None = None
-    trajectory_timesteps: list[torch.Tensor] | None = None
+    trajectory_timesteps: torch.Tensor | None = None
```
Comment (Rockdu) [inline python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py:340]: Double check this fix of trajectory_timesteps

File: python/sglang/multimodal_gen/runtime/models/schedulers/scheduling_flow_match_euler_discrete.py (hunk 3 above)
Comment (Rockdu) [reply python/sglang/multimodal_gen/runtime/models/schedulers/scheduling_flow_match_euler_discrete.py:455]: This is necessary considering multi-processing of different batches

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 5 above)
Comment (Rockdu) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:615]: This is necessary since rollout session data should be batch/session-level state

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 7]
```diff
@@ -130,6 +136,43 @@ def __init__(
         self._cached_num_steps = None
         self._is_warmed_up = False
 
+    def _maybe_prepare_rollout(self, batch: Req):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:143]: consider move to a dedicated mixin file for RL as a follow-up

**Review by mickqian — COMMENTED** [review]
excellent, nice and clean

**Review by mickqian — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 7 above)
Comment (Rockdu) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:143]: In the following PR denoising backpass we moved everything into RolloutDenoisingEnvMixin

File: python/sglang/multimodal_gen/test/run_suite.py [hunk 8]
```diff
@@ -41,6 +41,16 @@ def _discover_unit_tests() -> list[str]:
 
 SUITES = {
     # no GPU required; safe to run on any CPU-only runner
+    "unit": [
```
Comment (mickqian) [inline python/sglang/multimodal_gen/test/run_suite.py]: remove this since L54

Comment (Rockdu) [reply python/sglang/multimodal_gen/test/run_suite.py]: removed in latest commit
